daemon/oci_linux.go

```diff
@@ -316,12 +319,15 @@ func setNamespaces(daemon *Daemon, s *specs.Spec, c *container.Container) error
 			nsUser.Path = fmt.Sprintf("/proc/%d/ns/user", ic.State.GetPID())
 			setNamespace(s, nsUser)
 		}
-	} else if c.HostConfig.IpcMode.IsHost() {
+	case ipcMode.IsHost():
 		oci.RemoveNamespace(s, specs.LinuxNamespaceType("ipc"))
-	} else {
+	case ipcMode.IsPrivate(), ipcMode.IsShareable():
 		ns := specs.LinuxNamespace{Type: "ipc"}
 		setNamespace(s, ns)
+	default:
```

Comment:
In general, `""` is valid mode for per-container config, meaning "use the daemon default".

In this context, though, anything that is not `IsContainer()`, `IsHost()`, `IsPrivate()`, or `IsShareable()` is invalid. It can happen for the following reasons:
 - a new mode is added, but is not handled above;
 - somehow a mode string escaped the validation;
 - daemon default is left unset.
If any of these things happen, we have an error here.